[input] fallback to direct libinput mapping

oliver opened the pull request:
## Description
This is a possible fix for https://github.com/xbmc/xbmc/issues/16834 when using the GBM system.

This approach adds a fallback mapping table, which is used when XKB cannot map a key press to a Kodi keysym. In those cases the new fallback mapping table can directly map libinput keycodes to Kodi keysyms.
I think using the XKB mapping as first step is still a good idea, because it handles most of the keys well, including modifiers etc. The fallback table is only used for the keycodes that XKB cannot handle.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

## Motivation and context
Under GBM Kodi gets keyboard input from libinput, which works well. However, it then uses XKB to convert keycodes to (ultimately) Kodi keysyms. But XKB does not yet support many of the media keys (like "EPG" or "subtitle") [^1].

[^1]: previously I had written here that "XKB does not support extended keycodes"; but after discussion in this issue it turned out that I was wrong: XKB _does_ support "extended keycodes", but it does not know about many of the specific keys that we need in Kodi.

<!--
But XKB does not support "extended keycodes", i.e. keycodes above 255 (<https://who-t.blogspot.com/2021/01/auto-updating-xkb-for-new-kernel.html> states: "[4] X has an 8 bit keycode limit and that won't change until someone develops XKB2 with support for 32-bit keycodes, i.e. never.").

This means that Kodi currently cannot handle these extended keycodes at all, which affects keys like "EPG" or "subtitle".
-->

The current workarounds are tools like [`eventlircd`](https://wiki.libreelec.tv/configuration/ir-remotes) or `inputlircd`, or manipulating the kernel IR keymap to send other keycodes which are supported by XKB.

With this PR, XKB is still used for those keycodes that it currently supports. But the keycodes that are not supported by XKB yet are mapped using a new mapping table, which directly maps from libinput keycodes (i.e. evdev keycodes) to Kodi keysyms.

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

This PR is one of two possible solutions I implemented for issue #16834 (the alternative solution is at https://github.com/xbmc/xbmc/pull/26038). This PR works for some of the keys, and it is quite non-invasive. But it does not handle all keys, because Kodi does have "vkey" constants for all of the keys that a remote control can deliver. My understanding is that these vkey constants would have to be added to the `XBMCVKey` enum, which however is (according to docs) limited to 0xFF, meaning we probably cannot add constants for all remote control keys.

I have added a vkey constant for the "subtitle" key, because that one is important to me :-)

## How has this been tested?
I have tested this PR on an x86 system under Debian 12 (Bookworm), with Kodi running in Docker. I tested with a "One-For-All" universal remote control with two different USB DVB-T sticks:
- a "Dexatek Technology Ltd. DK DVB-T Dongle" (1d19:1101), which uses the `dvb_usb_rtl28xxu` driver
- a "Hauppauge Nova-T Stick 3" (2040:7070), which uses the `dvb_usb_dib0700` driver

In both cases the remote control used the RC-5 infrared protocol, and I used the same (hand-made) file in `/etc/rc_keymaps/` to map the IR codes to Linux evdev keycodes.

With the changes from the PR I could use all keys of my remote control that I had actually mapped (some keys don't emit an IR code at all in my current configuration).


<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## What is the effect on users?
With this PR it _should_ be possible to just plug in an infrared receiver and remote control (as long as it is supported by the Linux system) and use most of its keys in Kodi without any additional setup.

However, some keys (like "Language" or "Title") probably still won't work. I don't know whether that's a real problem, though.

<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

oliver committed:
[libinput] rename and simplify XBMCKeyForKeysym() method signature

oliver committed:
[libinput, keymaps] add support for KEY_SUBTITLE found on some remotes

oliver committed:
[libinput] reorganize code in CLibInputKeyboard::ProcessKey() a bit

There should be no functional changes in this commit.

oliver committed:
[libinput] add fallback mapping table, to directly map libinput scancodes to Kodi keys

XKB does not support "extended keycodes", i.e. keycodes > 255.
This affects e.g. the EPG and subtitle keys.
So even though libinput provides a keycode for these keypresses, they
cannot be mapped to a Kodi keysym when using the XKB system and the
xkbMap table.

Therefore a fallback table is added: if a libinput keycode cannot be
mapped with XKB, this fallback table directly maps some extended keycodes
to Kodi keysyms.

According to https://wayland.freedesktop.org/libinput/doc/latest/architecture.html#device-initialization
"libinput only supports evdev devices". Therefore the evdevKeycodeMap table
directly uses evdev keycode constants.

oliver committed:
[libinput] add mapping entries for more libinput keys

This mapping table was created by listing all keys from
https://www.kernel.org/doc/html/v6.11/userspace-api/media/rc/rc-tables.html
that are not handled by XKB yet, and then looking in
system/keymaps/keyboard.xml for an action that matches
to that input.

Since Kodi does not have key syms for all evdev keys, most
mapping entries are disabled here.

oliver committed:
[libinput] add debug log message when XKB keysym cannot be mapped

oliver committed:
[keymaps] handle dedicated Pause keyboard key

oliver committed:
[libinput, keymaps] handle dedicated Info keyboard key

oliver committed:
[libinput] handle some XKB variations of Menu key

garbear commented on xbmc/platform/linux/input/LibInputKeyboard.cpp:
```diff
@@ -433,31 +487,47 @@ void CLibInputKeyboard::ProcessKey(libinput_event_keyboard *e)
+  uint32_t limitedScancode = libinputKeycode;
+  if (limitedScancode > std::numeric_limits<unsigned char>::max())
+  {
+    // Kodi scancodes are limited to unsigned char, pretend the scancode is unknown on overflow
```
Can't we update Kodi to use uint32_t scancodes?

garbear commented:
I like the idea of adding more scancodes to LibInputKeyboard.cpp. Does the logic get simpler if you do this?

```patch
diff --git a/xbmc/input/keyboard/XBMC_keyboard.h b/xbmc/input/keyboard/XBMC_keyboard.h
index cb12461a80..a450dbcf7a 100644
--- a/xbmc/input/keyboard/XBMC_keyboard.h
+++ b/xbmc/input/keyboard/XBMC_keyboard.h
@@ -38,7 +38,7 @@
  */
 struct XBMC_keysym
 {
-  unsigned char scancode; /* hardware specific scancode */
+  uint32_t scancode; /* hardware specific scancode */
   XBMCKey sym; /* SDL virtual keysym */
   XBMCMod mod; /* current key modifiers */
   uint16_t unicode; /* translated character */
```

yol commented:
I don't completely understand this yet :thinking: So any 8 bit limitation that might exist would be from X11. xkb does not have this limitation. So it does not apply to gbm/wayland/etc. anyway.

For example, for the subtitle key, it is a valid keycode in the evdev map of xkb: https://gitlab.freedesktop.org/xkeyboard-config/xkeyboard-config/-/blob/master/keycodes/evdev?ref_type=heads#L343

As outlined in the blog post you referenced, this is then turned into a corresponding XKB keysym. This keysym is what we usually use to determine the internal XBMCKey. Why does this not work for the keys that you have? Are there XKB mappings missing? If so, isn't that a bug with xkb? I think we should avoid a second lookup table if it is possible.

oliver commented:
Good points... I'll have another look at XKB and see if I can get these extended keycodes mapped there. Probably not this week though :-)

oliver commented:
FYI, I'm still looking into the missing key mappings in XKB. That's probably caused by some missing `_EVDEVK()` entries in https://gitlab.freedesktop.org/xorg/proto/xorgproto/-/blob/master/include/X11/XF86keysym.h . So probably the next steps are:

- finding out why those entries were never added, and how to get them added;
- then getting https://gitlab.freedesktop.org/xkeyboard-config/xkeyboard-config/-/blob/master/symbols/inet regenerated, and getting a new release of xkeyboard-config project;
- and adding vkey constants for the missing keys to Kodi.

oliver commented:
Rebased to master now.

FYI, the PR for the changes that will (eventually) fix XKB is at https://gitlab.freedesktop.org/xorg/proto/xorgproto/-/merge_requests/93 .

garbear commented on xbmc/windowing/wayland/XkbcommonKeymap.cpp:
```diff
@@ -191,7 +193,7 @@ static const std::map<xkb_keycode_t, XBMCKey> XkbKeycodeXBMCMappings = {
     {XKB_KEY_XF86AudioPlay, XBMCK_PLAY},
     {XKB_KEY_XF86AudioRandomPlay, XBMCK_SHUFFLE}
-    // XBMCK_FASTFORWARD clashes with XBMCK_MEDIA_FASTFORWARD
+// XBMCK_FASTFORWARD clashes with XBMCK_MEDIA_FASTFORWARD
```
Sorry to nitpick, I saw no other issues in this PR. Can you undo this change, and ignore Jenkins if he wants it back for style?

garbear commented:
I am totally fine with this PR. It's a good improvement. Tables make for cheap code, so this brings benefit with little maintenance burden.

I'll merge once you fix the whitespace change I noted.

oliver commented:
Huh, I wasn't aware that suggestions from Jenkins could be ignored :-)

And yes, I would prefer to have the fix in XKB; but making user-visible improvements with simple code is a good tradeoff.

garbear commented:
@yol Any objection?

yol commented on xbmc/platform/linux/input/LibInputKeyboard.cpp:
```diff
@@ -151,6 +153,57 @@ constexpr auto xkbMap = make_map<xkb_keysym_t, XBMCKey>({
     // XBMCK_FASTFORWARD clashes with XBMCK_MEDIA_FASTFORWARD
 });
 
+/// Additional mappings directly from libinput (i.e. evdev) keycodes to Kodi keys
```
explanation why this is necessary would be nice in the comment (i.e., they were forgotten in xkb and we don't want to wait for the merge request)

oliver replied:
Good point; added.

HiassofT commented:
I tested this with a HP branded RC6 MCE remote on LibreELEC / RPi5 with a gpio IR receiver and noticed a few buttons didn't produce a XKB keysym:
"TV" / KEY_TV, "RADIO" / KEY_RADIO, "LIVE TV" / KEY_TUNER and "RECORDED TV" / KEY_PVR
```
debug <general>: ProcessKey: unable to map key with keycode 377 (0x179), XKB keysym 0 (0x0).
debug <general>: ProcessKey: unable to map key with keycode 385 (0x181), XKB keysym 0 (0x0).
debug <general>: ProcessKey: unable to map key with keycode 386 (0x182), XKB keysym 0 (0x0).
debug <general>: ProcessKey: unable to map key with keycode 366 (0x16e), XKB keysym 0 (0x0).
```

Those input event codes seem to be included in your XKB PR but are missing in the fallback table.

It would be nice if you could add all input event codes from your XKB PR to the fallback table so that this change works on distros with current libxkb (which likely will be around for a couple of years)

I also noticed some buttons produced a keysym but kodi didn't have a mapping for it. Most noticable the "BACK" button which sends a KEY_EXIT event code isn't mapped:
```
2025-01-26 13:17:27.732 T:1215    debug <general>: ProcessKey: unable to map key with keycode 174 (0xae), XKB keysym 269025110 (0x1008ff56).
```

"DVD MENU" / KEY_DVD, "MUSIC" / KEY_AUDIO, "PICTURES" / KEY_CAMERA, "VIDEOS" / KEY_VIDEO also miss a kodi mapping but are not as important IMHO 
```
debug <general>: ProcessKey: unable to map key with keycode 389 (0x185), XKB keysym 268964229 (0x10081185).
debug <general>: ProcessKey: unable to map key with keycode 392 (0x188), XKB keysym 268964232 (0x10081188).
debug <general>: ProcessKey: unable to map key with keycode 212 (0xd4), XKB keysym 269025167 (0x1008ff8f).
debug <general>: ProcessKey: unable to map key with keycode 393 (0x189), XKB keysym 269025159 (0x1008ff87).
```

It's probably best to care about the missing kodi mappings in a separate PR though.

oliver committed:
[libinput] handle some XKB variations of numeric keys

These XKB keys are mapped from the evdev KEY_NUMERIC_0 etc. keys.

oliver commented:
@HiassofT Thanks for testing!

_Summary:_ for most of the keys you mention (except for KEY_EXIT) I simply don't know yet what they should be mapped to in Kodi; and even if we decide on a mapping there are still technical limitations.

------

If I understand the code correctly, each input key needs to be mapped to some constant from https://github.com/xbmc/xbmc/blob/master/xbmc/input/keyboard/XBMC_keysym.h#L28 . And that list currently does not have constants for TV, Radio, Tuner, PVR, nor for DVD, Music/Audio, Pictures/Camera, or Video.<br>
Only for KEY_EXIT we might be lucky: we could map it to XBMCK_ESCAPE. What do you think?

So I guess we need to add several `XBMCK_` constants to XBMC_keysym.h, and also add matching `XBMCVK_` constants to https://github.com/xbmc/xbmc/blob/master/xbmc/input/keyboard/XBMC_vkeys.h#L24. But in the latter file there can only be at most 255 entries (according to the comment).

And then we need to define the exact "meaning" of these keys. https://www.kernel.org/doc/html/v6.11/userspace-api/media/rc/rc-tables.html might be helpful; and the meanings also have to fit with the actions from https://github.com/xbmc/xbmc/blob/master/system/keymaps/keyboard.xml .

Btw. do you have suggestions which Kodi actions should be invoked for your missing keys?

So, I would prefer to add these additional constants in a separate PR. Unless we decide to use a completely different approach, to avoid the 255-entry limit of XBMC_vkeys.h. I would prefer to get some opinion from the experienced Kodi devs on this, though.

oliver committed:
[libinput] handle XKB ChannelUp/Down keys as PageUp/Down

oliver committed:
[libinput] add (disabled) XKB mapping entries for some media keys

These media keys are actually supported by XKB, but Kodi does not have key
constant for these.

oliver commented:
I noticed that some keys from the new mapping table are actually supported by XKB, so now I added these to `xkbMap` instead. The remaining keys in `evdevKeycodeMap` are all still missing from XKB.

HiassofT commented:
I think it would be best to extend both XBMCK and XBMCVK with the missing codes, more or less adding the current IR remote button codes from https://github.com/xbmc/xbmc/blob/master/xbmc/input/keymaps/remote/IRTranslator.cpp#L162-L305 which don't have an equivalent in the current keyboard tables so the buttons/keys can be mapped to actions with keyboard.xml.

As for naming: I think it's better to use the XKB and/or linux input event names (eg TV, RADIO, ...) instead of the ones from IR translator.

I'd also prefer to have channelup/down as new XBMC(V)K values, remote.xml supports both channelup/down and pageup/down and mapping XKB_KEY_XF86ChannelUp/Down directly to XBMCK_PAGEUP/DOWN would remove the possibility to handle them separately. Same goes for KEY_EXIT, better make it a distinct XBMC(V)K as well than mapping to escape.

I haven't looked at how to best handle current remotes via keyboard.xml (that's likely a larger task) but as long as we don't loose any information in kodi's (lib)input code path and have the possibility to do all the mapping in keyboard.xml it should be a good starting point.

BTW: I'm happy to handle all that in separate follow-up PRs, just writing down my thoughts about where we should head to at the end

HiassofT commented:
Forgot to add: I'd also prefer to have a separate "ok" key (instead of mapping that to enter) in kodi, the MCE rremote has both a "OK" and a "ENTER" button: https://github.com/torvalds/linux/blob/master/drivers/media/rc/keymaps/rc-rc6-mce.c

HiassofT commented:
Thanks for the reminder about your other PR but I think it's better to focus on libinput as that will also work with non-rc-subsystem devices, eg bluetooth and other RF remotes with a USB dongle which can be hard to identify as remotes via udev.

garbear commented:
Don't be afraid to think outside the box. Like writing workarounds for the 8-bit issue, when you can instead change the box to use 32-bit. In 2025 is there a reason for the 255-entry limit of XBMC_vkeys.h?

garbear commented:
I had AI generate a PR, giving it only the text but not the diff. Here's what it pushed out: https://github.com/garbear/xbmc/commit/0f18ca2a82ab33729d13db294c74f32c9ed8981c . Anything there we can use in this PR?

Regarding the choice between the two PRs, the code here looks much more generic and reviewable, and I think it's preferable. Or maybe we do both approaches to maximize likelihood of input working.

bam80 commented:
> But XKB does not support "extended keycodes", i.e. keycodes above 255 (https://who-t.blogspot.com/2021/01/auto-updating-xkb-for-new-kernel.html states: "[4] X has an 8 bit keycode limit and that won't change until someone develops XKB2 with support for 32-bit keycodes, i.e. never.").

So the original article refers to X here, does it mean it still works without it, i.e. supported by `libxkbcommon`?

bam80 commented:
> But XKB does not support "extended keycodes", i.e. keycodes above 255 (https://who-t.blogspot.com/2021/01/auto-updating-xkb-for-new-kernel.html states: "[4] X has an 8 bit keycode limit and that won't change until someone develops XKB2 with support for 32-bit keycodes, i.e. never.").

So what it actually means is XKB _as a protocol_ (backed by X) doesn't support the "extended keycodes" indeed,
but XKB as a "system" (backed by `libxkbcommon`) does very well support them.

If that's true (correct me if I'm wrong), alternative of this PR and #26038 altogether is just:
- merge https://gitlab.freedesktop.org/xorg/proto/xorgproto/-/merge_requests/93 upstream
- add missing mapping here:
  https://github.com/xbmc/xbmc/blob/523ef087c101143e6dcf08c057c484a417993bfe/xbmc/platform/linux/input/LibInputKeyboard.cpp#L37-L39

@oliver what do you think?

bam80 commented:
> In the original description of this PR I had written this incorrectly, because I did not know better yet;

Thanks, would you mind to adjust the description in both of your PRs, then?
Because when I read it first time I got a wrong impression about the XKB ability (despite I have some background about that).

bam80 commented:
@oliver,
> -    the missing keys need to be added to XKB (by merging https://gitlab.freedesktop.org/xorg/proto/xorgproto/-/merge_requests/93),
> -    and then https://gitlab.freedesktop.org/xkeyboard-config/xkeyboard-config project needs to use the new XKB version, so that the new keys are added to the https://gitlab.freedesktop.org/xkeyboard-config/xkeyboard-config/-/blob/master/symbols/inet file,
> -   and I guess then we need to build Kodi against a new xkeyboard-config libxkbcommon version (?),

Considering https://gitlab.freedesktop.org/xorg/proto/xorgproto/-/merge_requests/93#note_2762262:
> meanwhile you can start to use directly the numeric value 0xnnnn of the keysyms in your keymaps and software.

, we don't have to wait for all these?

> -   and we need to add "vkey constants" for the new keys in Kodi, and also add these vkeys to the XBMCVKey enum,
> -   and we need to add mappings for these new keys to xkbMap.

I would rather add missing mappings for the existing vkeys first. All the other is really optional and could be done on demand.

bam80 commented:
> But to actually test the changes we also need to patch the XKB config files, e.g. in /usr/share/X11/xkb/symbols/inet in "autogenerated" section add the mappings for the new media keys.

What changes do you mean, is the patched XKB config needed even for direct `_EVDEVK` macro use?

oliver commented:
I'm not sure (and I cannot test right now); but IIRC without the changes to the `symbols/inet` file, XKB will not even map the keycodes to any keysyms, and `xkb_state_key_get_one_sym()` will return an error (i.e. `XKB_KEY_NoSymbol`). So in that case the contents of `xkbmap` would not even be relevant, because the processing would fail at an earlier step.

bam80 commented:
> Once we actually replace the numeric values in xkbmap with the constants from new XKB version, Kodi will only build against the brand new XKB version. Is that acceptable?

I think we would need to bump a minimum `libxkbcommon` version required in CMake.

garbear commented:
It looks like upstream is merged! Can we continue this PR?

garbear commented:
Make a new header and define all the new keys, e.g.

```c++
#if !defined(XKB_KEY_XF86OK)
#define XKB_KEY_XF86OK ...
#endif
```

Maybe put it in a `compat` folder.

Then use them like normal in LibInputKeyboard.cpp.

garbear commented:
Can you also include the keys from https://github.com/xbmc/xbmc/pull/16906 if they're not already in this PR?

HiassofT commented:
FYI: xkbcommon 1.11.0 was released this weekend
https://github.com/xkbcommon/libxkbcommon/releases/tag/xkbcommon-1.11.0

garbear commented:
@oliver We're doing a release of v22 Alpha 1 in the next few days or so. Do you want to get the additional keys into Alpha 1?

garbear commented:
Alpha 2 is now out! There's still no rush, we have a while until Alpha 3. I'm upgrading xkcbommon to the latest 1.12 release as well.

garbear commented:
@oliver I'm doing another pass of leftover work for Alpha 3, and came across this. I don't actually see any issues and wouldn't object to merging as-is. What do you think?

@kodiai please review

garbear commented:
Now that https://github.com/xbmc/xbmc/pull/27932 is in, where do we stand on this PR?

oliver commented:
Thanks for merging https://github.com/xbmc/xbmc/pull/27932! That covers most of the keys from this PR, with a better approach. So this PR is now obsolete, and I'm closing it.

I guess any discussion about further missing keys could happen in https://github.com/xbmc/xbmc/issues/26884 .
